### DISCUSSION TITLE
Improve filter searches performance by preventing the search from being called for each change in the filters

### DISCUSSION THREAD

**Discussion by RemiMatrod** [directus/directus#15780] [Feature Requests]
In the case where we are working on database tables with large amounts of data and a large amount of results displayed per page, each time the search is updated (by typing/removing a character in the search bar, or when updating the filter list), a new request is sent to the database.
This behaviour can cause massive slowdowns in the searching process, as we are waiting for the whole request to be treated by the database, and then are waiting for the page to display the results (the page becomes frozen until everything is shown, this can be problematic when the table display is set to show 500 lines per page).

Would it be possible to add an option in the search bar to prevent the search from being sent for each change done in the filters? This would let users be able to prepare their filters in advance before sending only one request with all the filters set, which can help with performance and user experience, and also limit the number of requests done to the database itself.

**Comment by rijkvanzanten** [comment]
I've also ran into this recently. @benhaynes could we brainstorm on what a "apply filters" button could look like in the context of the filtering UI?

**Reply by benhaynes** [reply]
Yup!!

**Reply by daawaan4x** [reply]
Just wanted to chime in, do queries with knex also get cancelled whenever the underlying request gets cancelled? I can't find any instances of `AbortController`s or any equivalent mechanisms present in the backend. In the frontend, I noticed by the way that the `useItems()` composable the built-in layouts are using appears to unintentionally throttle the cancellation too whenever the filter/search gets changed wherein the cancellation should happen outside the throttled function. Is this a bug? @rijkvanzanten 

https://github.com/directus/directus/blob/main/packages/shared/src/composables/use-items.ts

**Reply by rijkvanzanten** [reply]
> the filter/search gets changed wherein the cancellation should happen outside the throttled function. Is this a bug?

Could be! That being said, I think the only real scalable solution here is to use an "apply" button in the filtering UI, and drop the cancelling completely. It was a good idea in theory, but like you mentioned, the API doesn't cancel any internal operations when the network request is aborted, so it doesn't really alleviate API overhead on that front 👍🏻

**Reply by daawaan4x** [reply]
I also like the idea of having an "apply" button for the filters possibly together with the search bar listening to the enter key too to trigger refreshes, but I suggest it should be an optional feature only that can be configured individually for each layout for each collection (since searching in small tables with immediate results for QoL won't have any performance impact)

**Reply by daawaan4x** [reply]
on a related note, are there current plans on the roadmap to implement cancellations in the API? i thinking of that use case where the end-user has "applied" the filter resulting to a long-running query but forgot to change something in the filter so they'll "reapply" the filter again while the results haven't returned yet

edit: in addition, the implementation of cancellations cover a wider domain beyond this search performance issue though

**Reply by rijkvanzanten** [reply]
> I also like the idea of having an "apply" button for the filters possibly together with the search bar listening to the enter key too to trigger refreshes, but I suggest it should be an optional feature only that can be configured individually for each layout for each collection (since searching in small tables with immediate results for QoL won't have any performance impact)

I think we could do that smartly based on the complexity of the filter and the number of items in a collection. If you're only applying a single `field _eq x`, we could just do it debounced like we're doing it right now. As soon as you add more than n filters (we'll have to run some benchmarks) we can switch to an apply button 🙂

**Reply by rijkvanzanten** [reply]
> on a related note, are there current plans on the roadmap to implement cancellations in the API? i thinking of that use case where the end-user has "applied" the filter resulting to a long-running query but forgot to change something in the filter so they'll "reapply" the filter again while the results haven't returned yet
> 
> edit: in addition, the implementation of cancellations cover a wider domain beyond this search performance issue though

I don't think so. "Cancellations" in the API are fairly tricky, as the only real thing to cancel are the sequel queries, which could cause a bunch of side effects / corruption if those are cancelled mid-execution. I think the better solution to performance issues incurred by "spamming" large queries is improved parallelization and improved querying (f.e. queuing large requests to be handled sequentially and/or saving to a file (like an export) instead of trying to return in the same network request)

**Reply by daawaan4x** [reply]
> "Cancellations" in the API are fairly tricky, as the only real thing to cancel are the sequel queries, which could cause a bunch of side effects / corruption if those are cancelled mid-execution.

doesn't this only apply to requests that mutate state though, i think requests that only read data being cancelled mid-execution should be fine

**Reply by rijkvanzanten** [reply]
@daawaan4U _should_ being the keyword 😄 I have ran into some issues before where a cancelled read in a transaction causes the transaction to wait for a commit that never happens (as the request is cancelled), which in turn takes up one of the available connections. Easy to miss, but at a certain point the database just runs out of available connections and things get nasty real quick. Just saying that adding mid-query cancellation might add more complexity than we might think initially 🙂

**Reply by benhaynes** [reply]
In regards to the original discussion: I wasn't sure if we're talking about the SEARCH, FILTER, or BOTH. But, instead of adding the UI and extra step of "Apply", I am thinking of two options:

1. Smart Debouncing: Perhaps a debouncing strategy that changes (increases) based on the items per page count or total items in the collection? This doesn't really solve the issue but could help when typing into a `CONTAINS` field.
2. Instead of an "Apply" button, what if we just had the filters apply when the dropdown is closed? So you expand the filter menu, change things around... and when you blur/close the menu it applies the updates? This could be based on collection size too... so it only applies when you have more than X records.

**Reply by rijkvanzanten** [reply]
> what if we just had the filters apply when the dropdown is closed?

That could actually work for both! Only apply the filters when the searchbar/filter-box loses focus

**Comment by rijkvanzanten** [comment]
Heya! Thank you for taking the time to submit this request! 

It has been over **90 days**, and this discussion has not received at least **15 votes** from the community. This means that we don't feel like there's enough community interest to warrant further R&D into this topic at this time. :ice_cube:

This request will now be closed to keep our discussions tidy. Please reach out if you have any questions!

_For more information, see our [Feature Request Process](https://docs.directus.io/contributing/feature-request-process.html)._
